src/intfuncs.c

```diff
@@ -617,6 +617,211 @@ Obj FuncINT_STRING ( Obj self, Obj string )
+* use the NLOC_FUNC and FEXS_FUNC fields, which we alias as
+* MASK_BITFIELD_FUNC and OFFSET_BITFIELD_FUNC.
+*
+* For fields of size 1 we also offer Boolean setters and getters which
+* accept and return True for 1 and False for 0. This makes for much
+* nicer code on the GAP side.
+*
+*/
+
+
+static inline UInt MASK_BITFIELD_FUNC(Obj func)
+{
+  return NLOC_FUNC(func);
+}
+
+static inline void SET_MASK_BITFIELD_FUNC(Obj func, UInt mask)
+{
+  SET_NLOC_FUNC(func, mask);
+}
+
+static inline UInt OFFSET_BITFIELD_FUNC(Obj func)
+{
+  GAP_ASSERT(IS_INTOBJ(FEXS_FUNC(func)));
+  return INT_INTOBJ(FEXS_FUNC(func));
+}
+
+static inline void SET_OFFFSET_BITFIELD_FUNC(Obj func, UInt offset)
+{
+  return SET_FEXS_FUNC(func, INTOBJ_INT(offset));
+		    
+}
+
+static Obj DoFieldGetter(Obj self, Obj data)
+{
+  UInt mask = MASK_BITFIELD_FUNC(self);
+  UInt offset = OFFSET_BITFIELD_FUNC(self);
+    if (!IS_INTOBJ(data))
+        ErrorMayQuit("Field getter: argument must be small integer", 0, 0);
+    UInt x = INT_INTOBJ(data);
+    return INTOBJ_INT((x & mask)>>offset);
+}
+
+static Obj DoFieldSetter(Obj self, Obj data, Obj val)
+{
+  UInt mask = MASK_BITFIELD_FUNC(self);
+  UInt offset = OFFSET_BITFIELD_FUNC(self);
+    if (!ARE_INTOBJS(data, val))
+        ErrorMayQuit("Field Setter: both arguments must be small integers", 0,
+                     0);
+    UInt x = INT_INTOBJ(data);
+    UInt y = INT_INTOBJ(val);
+    return INTOBJ_INT((x & ~mask) | (y << offset));
+}
+
+static Obj DoBooleanFieldGetter(Obj self, Obj data)
+{
+  UInt mask = MASK_BITFIELD_FUNC(self);
+    if (!IS_INTOBJ(data))
+        ErrorMayQuit("Boolean Field getter: argument must be small integer", 0, 0);
+    UInt x = INT_INTOBJ(data);
+    return (x & mask) ? True : False;
+}
+
+static Obj DoBooleanFieldSetter(Obj self, Obj data, Obj val)
+{
+  UInt mask = MASK_BITFIELD_FUNC(self);
+    if (!IS_INTOBJ(data))
+        ErrorMayQuit("Boolean Field Setter: data must be small integer", 0,
+                     0);    
+    UInt x = INT_INTOBJ(data);
+    if (val == True)
+      x |= mask;
+    else if (val == False)
+      x &= ~mask;
+    else
+      ErrorMayQuit("Boolean Field Setter: value must be true or false", 0, 0);
+    return INTOBJ_INT(x);
+}
+
+
+static Obj FuncBUILD_BITFIELDS(Obj self, Obj args)
+{
+  GAP_ASSERT(IS_PLIST(args));
+  GAP_ASSERT(LEN_PLIST(args) >= 1  && ELM_PLIST(args,1));
+  Obj widths = ELM_PLIST(args,1);
+  if (!IS_LIST(widths))
+    ErrorMayQuit("Fields builder: first argument must be list of widths", 0,
+                     0);
+  UInt nfields = LEN_LIST(widths);
+  if (LEN_PLIST(args) != nfields + 1)
+    ErrorMayQuit("Fields builder: number of values must match number of widths", 0, 0);
+  UInt x = 0;
+  UInt i;
+  for (i = nfields; i > 0; i--) {
+    GAP_ASSERT(ISB_LIST(widths, i));
+    Obj y = ELM_LIST(widths,i);
+    GAP_ASSERT(IS_INTOBJ(y));
+    x <<= INT_INTOBJ(y);
+    GAP_ASSERT(ELM_PLIST(args, i+1));
```

Comment:
That `GAP_ASSERT` is somewhat redundant, as the `IS_INTOBJ` test later on will simply fail in this case. But of course it doesn't hurt to have here.